tests/playwright-test/test-server-connection.spec.ts

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { test as baseTest, expect } from './ui-mode-fixtures';
+
+import { TestServerConnection } from '../../packages/playwright/src/isomorphic/testServerConnection';
+
+class TestServerConnectionUnderTest extends TestServerConnection {
+  events: [string, any][] = [];
+
+  constructor(wsUrl: string) {
+    super(wsUrl);
+    this.onTestFilesChanged(params => this.events.push(['testFilesChanged', params]));
+    this.onStdio(params => this.events.push(['stdio', params]));
+    this.onLoadTraceRequested(params => this.events.push(['loadTraceRequested', params]));
+    this.onReport(params => this.events.push(['report', params]));
+  }
+}
+
+const test = baseTest.extend<{ testServerConnection: TestServerConnectionUnderTest }>({
+  testServerConnection: async ({ startCLICommand }, use) => {
+    const testServerProcess = await startCLICommand({}, 'test-server');
+
+    await testServerProcess.waitForOutput('Listening on');
+    const line = testServerProcess.output.split('\n').find(l => l.includes('Listening on'));
+    const wsEndpoint = line!.split(' ')[2];
+
+    await use(new TestServerConnectionUnderTest(wsEndpoint));
+
+    await testServerProcess.kill();
+  }
+});
+
+test('file watching', async ({ testServerConnection, writeFiles }, testInfo) => {
+  await writeFiles({
+    'utils.ts': `
+      export const expected = 42;
+      `,
+    'a.test.ts': `
+      import { test } from '@playwright/test';
+      import { expected } from "./utils";
+      test('foo', () => {
+        expect(123).toBe(expected);
+      });
+      `,
+  });
+
+  const tests = await testServerConnection.listTests({});
+  expect(tests.report.map(e => e.method)).toEqual(['onConfigure', 'onProject', 'onBegin', 'onEnd']);
+
+  await testServerConnection.watch({ fileNames: ['a.test.ts'] });
+
+  await writeFiles({
+    'utils.ts': `
+      export const expected = 123;
+      `,
+  });
+
+  await expect.poll(() => testServerConnection.events).toHaveLength(1);
+  expect(testServerConnection.events).toEqual([
+    ['testFilesChanged', { testFiles: [testInfo.outputPath('a.test.ts')] }]
+  ]);
+});
+
+test('test run events', async ({ testServerConnection, writeFiles }) => {
```

Comment:
Perhaps it would make more sense to test the end-user behavior? While I do think such a test makes sense for a public API that should preserve backwards compatibility, test server interface is not public yet. WDYT?

Reply:
This API might not be public, but we still need to preserve backwards compatibility so we don't break the VS Code extension. That's the main value in this test suite I think.

Reply:
`report` also is one of the events of `TestServerInterface`, but I agree that if its contents exclusively contain reporter output, then it doesn't make sense to test it.

In https://github.com/microsoft/playwright/pull/31842/commits/9ee6994979b3332bba881f265c70131e1e158491, i've removed the `report` event and rewrote this test to cover STDIO interception instead.